src/components/Trade/hooks/useSwapper/useSwapper.ts

```diff
@@ -50,157 +54,238 @@ export enum TRADE_ERRORS {
   const [debounceObj, setDebounceObj] = useState<{ cancel: () => void }>()
 
-  useEffect(() => {
-    actionRef.current = action
-  }, [action])
-
   const getDefaultPair = useCallback(() => {
     const swapper = swapperManager.getSwapper(bestSwapperType)
     return swapper.getDefaultPair()
   }, [swapperManager, bestSwapperType])
 
+  const buildQuoteTx = async ({
+    wallet,
+    sellAsset,
+    buyAsset,
+    amount
+  }: {
+    wallet: HDWallet
+    sellAsset: Asset
+    buyAsset: Asset
+    amount: string
+  }): Promise<Quote<ChainTypes, SwapperType> | undefined> => {
+    let result
+    try {
+      const swapper = swapperManager.getSwapper(bestSwapperType)
+      result = await swapper?.buildQuoteTx({
+        input: {
+          sellAmount: toBaseUnit(amount, sellAsset.precision),
+          sellAsset,
+          buyAsset,
+          sellAssetAccountId: '0', // TODO: remove hard coded accountId
+          buyAssetAccountId: '0', // TODO: remove hard coded accountId
+          slippage: trade?.slippage?.toString(),
+          priceImpact: quote?.priceImpact,
+          sendMax: false // TODO: implement sendMax
+        },
+        wallet
+      })
+    } catch (err) {
+      console.error(`TradeProvider - buildTransaction error: ${err}`)
+    }
+    if (result?.success) {
+      setFees(result, sellAsset)
+      setValue('quote', result)
+    } else {
+      // TODO: (ryankk) fix errors to reflect correct trade attribute
+      setError('useSwapper', { message: TRADE_ERRORS.INSUFFICIENT_FUNDS })
+    }
+    return result
+  }
+
+  const executeQuote = async ({
+    wallet
+  }: {
+    wallet: HDWallet
+  }): Promise<ExecQuoteOutput | undefined> => {
+    let result
+    try {
+      const swapper = swapperManager.getSwapper(bestSwapperType)
+      result = await swapper.executeQuote({ quote, wallet })
+    } catch (err) {
+      setError('useSwapper', { message: TRADE_ERRORS.NO_LIQUIDITY })
+      console.error(`TradeProvider - executeQuote error: ${err}`) // eslint-disable-line no-console
+    }
+    return result
+  }
+
   const getQuoteFromSwapper = async <C extends ChainTypes, S extends SwapperType>({
     amount,
     sellAsset,
     buyAsset,
+    action,
     onFinish
-  }: GetQuote<C, S>) => {
+  }: GetQuoteFromSwapper<C, S>) => {
     if (debounceObj?.cancel) debounceObj.cancel()
     clearErrors()
     const quoteDebounce = debounce(async () => {
-      try {
-        const swapper = swapperManager.getSwapper(bestSwapperType)
-        let convertedAmount = amount
-        const isFiat = Object.keys(amount)[0].includes('fiat')
-        if (isFiat) {
-          const rate = await swapper.getUsdRate({
+      if (isComponentMounted.current) {
+        try {
+          const swapper = swapperManager.getSwapper(bestSwapperType)
+          let convertedAmount =
+            action === TradeActions.BUY ? { buyAmount: amount } : { sellAmount: amount }
+
+          const isFiat = action === TradeActions.FIAT
+          if (isFiat) {
```

Comment:
if you don't use `isFiat` later in the function
```suggestion
          if (action === TradeActions.FIAT) {
```